File: Sources/AdbShell/Services/ApkBadgingParser.swift
import Foundation

/// Разбор `aapt2 dump badging <apk>` — тот же вывод, что уже используется
/// IconService для извлечения пути к иконке, здесь вытаскиваем более полную сводку.
enum ApkBadgingParser {
    static func parse(_ output: String) -> ApkManifestInfo {
        var packageName: String?
        var versionName: String?
        var versionCode: String?
        var minSdk: String?
        var targetSdk: String?
        var applicationLabel: String?
        var permissions: [String] = []

        for line in output.split(separator: "\n") {
            if line.hasPrefix("package:") {
                packageName = attribute(in: line, name: "name")
                versionCode = attribute(in: line, name: "versionCode")
                versionName = attribute(in: line, name: "versionName")
            } else if line.hasPrefix("sdkVersion:") {
                minSdk = quotedValue(in: line)
            } else if line.hasPrefix("targetSdkVersion:") {
                targetSdk = quotedValue(in: line)
            } else if line.hasPrefix("application-label:") {
                applicationLabel = quotedValue(in: line)
            } else if line.hasPrefix("uses-permission:") {
                if let name = attribute(in: line, name: "name") {
                    permissions.append(name)
                }
            }
        }

        return ApkManifestInfo(
            packageName: packageName,
            versionName: versionName,
            versionCode: versionCode,
            minSdk: minSdk,
            targetSdk: targetSdk,
            applicationLabel: applicationLabel,
            permissions: permissions,
            rawBadging: output
        )
    }

    /// Значение вида `name='значение'` в произвольном месте строки.
    private static func attribute(in line: Substring, name: String) -> String? {
        guard let range = line.range(of: "\(name)='") else { return nil }
        let rest = line[range.upperBound...]
        guard let end = rest.firstIndex(of: "'") else { return nil }
        return String(rest[rest.startIndex..<end])
    }

    /// Значение вида `label:'значение'` сразу после первого двоеточия строки.
    private static func quotedValue(in line: Substring) -> String? {
        guard let colon = line.firstIndex(of: ":") else { return nil }
        let rest = line[line.index(after: colon)...]
        guard let firstQuote = rest.firstIndex(of: "'") else { return nil }
        let afterQuote = rest[rest.index(after: firstQuote)...]
        guard let end = afterQuote.firstIndex(of: "'") else { return nil }
        return String(afterQuote[afterQuote.startIndex..<end])
    }
}

File: desktop/src/main/adb/parsers/ApkBadgingParser.ts
// Порт Sources/AdbShell/Services/ApkBadgingParser.swift — разбор вывода
// `aapt2 dump badging <apk>` (тот же способ, что использует официальный
// Android Gradle Plugin для чтения манифеста локального .apk без установки
// на устройство).

import { ApkManifestInfo } from '../types/ApkManifestInfo';

export function parseApkBadging(output: string): ApkManifestInfo {
  let packageName: string | undefined;
  let versionName: string | undefined;
  let versionCode: string | undefined;
  let minSdk: string | undefined;
  let targetSdk: string | undefined;
  let applicationLabel: string | undefined;
  const permissions: string[] = [];

  for (const line of output.split('\n')) {
    if (line.startsWith('package:')) {
      packageName = attribute(line, 'name');
      versionCode = attribute(line, 'versionCode');
      versionName = attribute(line, 'versionName');
    } else if (line.startsWith('sdkVersion:')) {
      minSdk = quotedValue(line);
    } else if (line.startsWith('targetSdkVersion:')) {
      targetSdk = quotedValue(line);
    } else if (line.startsWith('application-label:')) {
      applicationLabel = quotedValue(line);
    } else if (line.startsWith('uses-permission:')) {
      const name = attribute(line, 'name');
      if (name !== undefined) permissions.push(name);
    }
  }

  return { packageName, versionName, versionCode, minSdk, targetSdk, applicationLabel, permissions, rawBadging: output };
}

/** Значение вида name='значение' в произвольном месте строки. */
function attribute(line: string, name: string): string | undefined {
  const marker = `${name}='`;
  const start = line.indexOf(marker);
  if (start === -1) return undefined;
  const rest = line.slice(start + marker.length);
  const end = rest.indexOf("'");
  if (end === -1) return undefined;
  return rest.slice(0, end);
}

/** Значение вида label:'значение' сразу после первого двоеточия строки. */
function quotedValue(line: string): string | undefined {
  const colon = line.indexOf(':');
  if (colon === -1) return undefined;
  const rest = line.slice(colon + 1);
  const firstQuote = rest.indexOf("'");
  if (firstQuote === -1) return undefined;
  const afterQuote = rest.slice(firstQuote + 1);
  const end = afterQuote.indexOf("'");
  if (end === -1) return undefined;
  return afterQuote.slice(0, end);
}
